packages/next-server/lib/router-context.ts

```diff
@@ -1,3 +1,6 @@
 import * as React from 'react'
+import Router from './router/router'
 
-export const RouterContext: React.Context<any> = React.createContext(null)
+export const RouterContext: React.Context<Router> = React.createContext(
```

Comment:
If you make this `BaseRouter` instead of `Router`, you won't have to assert `as any` [here](https://github.com/zeit/next.js/pull/7768/files#diff-92773cb42254dca07ea21eddbc04f922R332)

Reply:
It's not possible to use the base router in context, because we these types to propagate it to client types. The base router is missing methods what could a client use.

Reply:
Also, if we apply #7797 (to get generic `BaseRouter`), change this line to `React.Context<BaseRouter<any>>`

and then change `useRouter` to

```
export function useRouter<P>(): BaseRouter<P> {
  return React.useContext(RouterContext)
}
```

we'll be able to get query types through `useRouter`

Reply:
@huv1k you're talking about everything added [here](https://github.com/zeit/next.js/blob/3e51ddb8af55b6438aa3aeb382081b9a1c86f325/packages/next/client/router.ts#L19)?

Reply:
Yes

Reply:
It appears to me that specifying the type of the context as `Router` is incorrect, then, as it can be either `ServerRouter` or `PublicRouterInstance`. As it stands, the return type of `useRouter` is `Router` - which I don't think is intended (as that misses the same extra fields).

Reply:
@tills13 @huv1k I already have a fix for this, I'll push it soon, neither `BaseRouter` nor `Router` should be here because the real router that is added to the context is an instance of `PublicRouterInstance`.

Reply:
Server-side a `ServerRouter` is used in this context, though, no? 

https://github.com/zeit/next.js/pull/7768/files#diff-92773cb42254dca07ea21eddbc04f922R332

`router` is a `ServerRouter`.

Reply:
`ServerRouter` is also `PublicRouterInstance`, is was implementing `BaseRouter` but that was just a mistake.